// @celo/utils/lib/async.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sleep = exports.selectiveRetryAsyncWithBackOff = exports.retryAsyncWithBackOff = exports.retryAsync = exports.concurrentValuesMap = exports.concurrentMap = void 0;
// Exports moved to @celo/base, forwarding them
// here for backwards compatibility
var async_1 = require("@celo/base/lib/async");
Object.defineProperty(exports, "concurrentMap", { enumerable: true, get: function () { return async_1.concurrentMap; } });
Object.defineProperty(exports, "concurrentValuesMap", { enumerable: true, get: function () { return async_1.concurrentValuesMap; } });
Object.defineProperty(exports, "retryAsync", { enumerable: true, get: function () { return async_1.retryAsync; } });
Object.defineProperty(exports, "retryAsyncWithBackOff", { enumerable: true, get: function () { return async_1.retryAsyncWithBackOff; } });
Object.defineProperty(exports, "selectiveRetryAsyncWithBackOff", { enumerable: true, get: function () { return async_1.selectiveRetryAsyncWithBackOff; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return async_1.sleep; } });
