// @celo/contractkit/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityMetadataWrapper = exports.Token = exports.StableToken = exports.CeloTokens = exports.RegisteredContracts = exports.CeloContract = exports.AllContracts = exports.REGISTRY_CONTRACT_ADDRESS = exports.NULL_ADDRESS = void 0;
var address_1 = require("@celo/base/lib/address");
Object.defineProperty(exports, "NULL_ADDRESS", { enumerable: true, get: function () { return address_1.NULL_ADDRESS; } });
var address_registry_1 = require("./address-registry");
Object.defineProperty(exports, "REGISTRY_CONTRACT_ADDRESS", { enumerable: true, get: function () { return address_registry_1.REGISTRY_CONTRACT_ADDRESS; } });
var base_1 = require("./base");
Object.defineProperty(exports, "AllContracts", { enumerable: true, get: function () { return base_1.AllContracts; } });
Object.defineProperty(exports, "CeloContract", { enumerable: true, get: function () { return base_1.CeloContract; } });
Object.defineProperty(exports, "RegisteredContracts", { enumerable: true, get: function () { return base_1.RegisteredContracts; } });
var celo_tokens_1 = require("./celo-tokens");
Object.defineProperty(exports, "CeloTokens", { enumerable: true, get: function () { return celo_tokens_1.CeloTokens; } });
Object.defineProperty(exports, "StableToken", { enumerable: true, get: function () { return celo_tokens_1.StableToken; } });
Object.defineProperty(exports, "Token", { enumerable: true, get: function () { return celo_tokens_1.Token; } });
var identity_1 = require("./identity");
Object.defineProperty(exports, "IdentityMetadataWrapper", { enumerable: true, get: function () { return identity_1.IdentityMetadataWrapper; } });
__exportStar(require("./kit"), exports);
__exportStar(require("./proxy"), exports);
